### PR TITLE
feat: add more performant masonry grid #doNotMerge

### PR DESCRIPTION
**Pull request by gkartalis**
This PR resolves [PHIRE-3] 

### PR Checklist

- [ ] I have tested my changes on **iOS** and **Android**.
- [ ] I hid my changes behind a **[feature flag]**, or they don't need one.
- [ ] I have included **screenshots** or **videos**, or I have not changed the UI.
- [ ] I have added **tests**, or my changes don't require any.
- [ ] I added an **[app state migration]**, or my changes do not require one.
- [ ] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [ ] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

[PHIRE-3]: https://artsyproduct.atlassian.net/browse/PHIRE-3?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx [hunk 1]
```diff
@@ -193,8 +240,88 @@ const ArtistArtworksContainer: React.FC<ArtworksGridProps & ArtistArtworksContai
...
+          viewAreaCoveragePercentThreshold: 20,
+        }}
+        onViewableItemsChanged={onViewableItemsChanged}
+        contentContainerStyle={{
+          paddingHorizontal: space(2),
+          paddingBottom: space(4),
+          backgroundColor: "purple",
+        }}
+        // onEndReached={() => console.warn("onEndReached")}
+        // onEndReachedThreshold={0.1}
```
Comment (gkartalis) [inline src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx:320]: **onEndReached prop: (generally the correct way of fetching more data on an infinite scroll list)** 

If wrapped with Flex the onEndReached is getting called on the **first** render only.

if wrapped with Flex with 1 and another one with the height (apparently Masonry really likes being wrapped with a [**fixed** height](https://shopify.github.io/flash-list/docs/known-issues#1-flashlists-rendered-size-is-not-usable-warning) parent ) the onEndreached is getting called correctly **BUT** breaks the scrolling functionality of the sticky tabs.

File: src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx [hunk 2]
```diff
@@ -193,8 +240,88 @@ const ArtistArtworksContainer: React.FC<ArtworksGridProps & ArtistArtworksContai
...
+        }}
+        // onLoad={({ elapsedTimeInMs }) => {
+        //   console.log("List Load Time", elapsedTimeInMs)
+        // }}
+        viewabilityConfig={{
+          // itemVisiblePercentThreshold: 50,
+          waitForInteraction: true,
+          viewAreaCoveragePercentThreshold: 20,
+        }}
+        onViewableItemsChanged={onViewableItemsChanged}
```
Comment (gkartalis) [inline src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx:313]: **`onViewableItemsChanged` + `viewabilityConfig`** workaround:

tried the following workaround:

```js
viewabilityConfig={{
  waitForInteraction: true,
  viewAreaCoveragePercentThreshold: 20, // played around with different values here
}}

onViewableItemsChanged={
  useCallback(

({ changed }) => {


// check if the last item of each page is visible to fetch more
// artworksPageSize is basically :
// const artworks = extractNodes(artist.artworks)
// const artworksLength = artworks?.length ?? 0
// const artworksPageSize = artworksLength / 10
// in order to have a hacky way of tracking the page to fetch more items when 
// we need to in order to have an infinite scroll experience

const isEnd = changed.some(
(obj) => obj.index === PAGE_SIZE * artworksPageSize - 1 && obj.isViewable
)



relay.loadMore(PAGE_SIZE)

}

},

[artworksPageSize]

)
}
```


The issue with the above was that the `onViewableItemsChanged` callback is not aware of any state changes / variable changes within the component. 

This results in artworksPageSize getting incremented BUT within the context of `onViewableItemsChanged` it is always 1.

File: src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx [hunk 3]
```diff
@@ -193,8 +240,88 @@ const ArtistArtworksContainer: React.FC<ArtworksGridProps & ArtistArtworksContai
...
+        contentContainerStyle={{
+          paddingHorizontal: space(2),
+          paddingBottom: space(4),
+          backgroundColor: "purple",
+        }}
+        // onEndReached={() => console.warn("onEndReached")}
+        // onEndReachedThreshold={0.1}
+      />
```
Comment (gkartalis) [inline src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx:321]: Workaround with ListFooterComponent:

Having a 'load more' button on the end of the list to fetch more data solves our issues but breaks the infinite scroll experience 😅

**Comment by gkartalis** [comment]
Another approach would be to add Masonry to the collapsible-tabs library that we use but I highly doubt that it would work due to already having some layout issues with flashlist as a [known issue](https://github.com/PedroBern/react-native-collapsible-tab-view/issues/335) for FlashList (<MasonryFlashList is two flashlists under the hood).

In order to also add Masonry to the collapsible tabs we need to do something similar to [this](https://github.com/PedroBern/react-native-collapsible-tab-view/blob/main/src/FlashList.tsx) but the issue is that Flashlist and MasonryFlashList onScroll events work a bit differently (have already opened an [issue](https://github.com/Shopify/flash-list/issues/689) about that on flashlist repo).

And that makes me believe that this won't work either way.

**Comment by gkartalis** [comment]
Feel free to review or play around with it and let me know if I missed something. Not ready but open for feedback eitherway on my comments mostly.

**Review by araujobarret — COMMENTED** [review]
Thanks for all the exploration here first of all!

Seeing how many issues/hacks are necessary I don't feel super confident in using it, might become more of a problem than a helper.
Another downside that limits the usage of the Masonry is to need of having a parent View with a fixed height, not all the time we will have this value.

**Comment by gkartalis** [comment]
Waiting for design feedback to see how we proceed with that. Will close this till then and will reopen a new pr once we have an actual guideline

**Pull request closed without merging**